rest_api/test/test_rest_api.py

```diff
@@ -208,226 +123,307 @@ def populated_client(client: TestClient):
+    # containing only one label
+    args, _ = MockDocumentStore.mocker.write_labels.call_args
+    labels = args[0]
+    assert len(labels) == 1
+    # Ensure all the items that were in `feedback` are part of the stored label
+    label = labels[0].to_dict()
+    for k, v in feedback.items():
+        assert label[k] == v
+
+
+def test_write_feedback_without_id(client, feedback):
+    feedback = deepcopy(feedback)
     del feedback["id"]
-    response = populated_client.post(url="/feedback", json=feedback)
+    response = client.post(url="/feedback", json=feedback)
     assert 200 == response.status_code
-    assert api_document_store.get_label_count() == 1
-
-    label: Label = api_document_store.get_all_labels()[0]
-    label_values = label.to_dict()
-    for actual_item, expected_item in [(label_values[key], value) for key, value in FEEDBACK.items() if key != "id"]:
-        assert actual_item == expected_item
-
-
-def test_get_feedback(populated_client_with_feedback: TestClient):
-    response = populated_client_with_feedback.get(url="/feedback")
+    # Ensure `write_labels` was called on the Document Store instance passing a list
+    # containing only one label
+    args, _ = MockDocumentStore.mocker.write_labels.call_args
+    labels = args[0]
+    assert len(labels) == 1
+    # Ensure the `id` was automatically set before storing the label
+    label = labels[0].to_dict()
+    assert label["id"]
+
+
+def test_get_feedback(client):
+    response = client.get("/feedback")
     assert response.status_code == 200
-    json_response = response.json()
-    for response_item, expected_item in [(json_response[0][key], value) for key, value in FEEDBACK.items()]:
-        assert response_item == expected_item
+    MockDocumentStore.mocker.get_all_labels.assert_called_once()
 
 
-def test_delete_feedback(populated_client_with_feedback: TestClient, api_document_store: BaseDocumentStore):
-    response = populated_client_with_feedback.delete(url="/feedback")
-    assert 200 == response.status_code
-    assert api_document_store.get_label_count() == 0
+def test_delete_feedback(client, monkeypatch, feedback):
+    # This label contains `origin=user-feedback` and should be deleted
+    label_to_delete = Label.from_dict(feedback)
+    # This other label has a different origin and should NOT be deleted
+    label_to_keep = Label.from_dict(feedback)
+    label_to_keep.id = "42"
+    label_to_keep.origin = "not-from-api"
 
+    # Patch the Document Store so it returns the 2 labels above
+    def get_all_labels(*args, **kwargs):
+        return [label_to_delete, label_to_keep]
 
-def test_do_not_delete_gold_labels(populated_client_with_feedback: TestClient, api_document_store: BaseDocumentStore):
-    feedback = deepcopy(FEEDBACK)
-    feedback["id"] = "456"
-    feedback["origin"] = "gold-label"
-    api_document_store.write_labels([feedback])
+    monkeypatch.setattr(MockDocumentStore, "get_all_labels", get_all_labels)
```

Comment:
Why we need this? We're only checking what the mock was called with, not that is actually deletes labels... or have I misunderstood this test?

Reply:
The controller has some logic to filter out labels, we want to check that the final call to `delete` is performed with the right subset of labels

Reply:
Oh sorry I see it now! Sorry, dumb question :sweat_smile: